### DISCUSSION TITLE
Configure htlc_max via plugin or API

### DISCUSSION THREAD

**Discussion by NCrashed** [ACINQ/eclair#2213] [Developers]
Hi, I am an operator of routing Eclair node. There is a common practice to set higher fees when channel is depleted. That helps to reduce failing payment rate as wallets tend to select routes with lower fees. But, the approach conflicts with my marketing strategy to sustain low fees on my node. Dynamic fees skew statistics on explorers services.

There is also another way to signal wallets that the channel is not suited for payment. It is setting max_htlc parameter dynamically. The advantage of the method is that I can set constant fees and get encourage natural rebalancing simultaneously. 

I want to implement a plugin for Eclair to set max_htlc dynamically, but I struggle to find part of sources that I should dig in. Could someone help me by pointing where to start?

**Comment by pm47** [comment]
> There is also another way to signal wallets that the channel is not suited for payment

The "proper" way of signalling that a channel shouldn't be used would actually be to set the `disable` bit of `channel_update`. This is what we do at channel creation when you are fundee, the channel will only be enabled after your side goes above reserve (after you receive a few payments). That's a similar use case as the one you are describing. But apart from that, this bit is managed by eclair and used mostly in relation to connections/disconnections, so whatever you do with a plugin would be overridden.

> I want to implement a plugin for Eclair to set max_htlc dynamically, but I struggle to find part of sources that I should dig in. Could someone help me where to start?

You can't customize this value, as per the [protocol spec](https://github.com/lightning/bolts/blob/master/07-routing-gossip.md#the-channel_update-message) for privacy reasons:
> Note that the htlc_maximum_msat field is static in the current protocol over the life of the channel: it is not designed to be indicative of real-time channel capacity in each direction, which would be both a massive data leak and uselessly spam the network.

Let's say you change your mind and choose to implement it with custom fees, here is how you would do it:
1) listen to the [`AvailableBalanceChanged`](https://github.com/ACINQ/eclair/blob/fd849095854547dfcf0af5e80aa906fa36cdbbf4/eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelEvents.scala#L69) event
2) upon receiving a `AvailableBalanceChanged`, look at `event.commitments.availableBalanceForSend` and maybe send a `CMD_UPDATE_RELAY_FEE` to the channel

**Reply by NCrashed** [reply]
@pm47 thanks for clearing that out! So, could I use disable/enable mechanism to mark depleted channels?

**Reply by pm47** [reply]
No, as I mentioned, this bit is managed by eclair, so your only option today is to use the fees.

**Comment by DerEwige** [comment]
@pm47  

> Note that the htlc_maximum_msat field is static in the current protocol over the life of the channel: it is not designed to be indicative of real-time channel capacity in each direction, which would be both a massive data leak and uselessly spam the network.

If I were to only update the max_htlc when I update the fee anyway, that would not increase the traffic on the network, right? The value is present in each update anyway afaik?

If I am ok with my channel state being leaked out to others, wouldn’t this massively help in route finding?

**Reply by t-bast** [reply]
> If I were to only update the max_htlc when I update the fee anyway, that would not increase the traffic on the network, right? The value is present in each update anyway afaik?

Yes, but we don't want to encourage you to create `channel_update`s more frequently: quite the opposite actually, otherwise your `channel_update`s may be identified as spam and ignored by other nodes (cln already drops a non-negligible percentage of the `channel_update`s they receive).

> If I am ok with my channel state being leaked out to others, wouldn’t this massively help in route finding?

If you're ok with revealing channel state information, then you should use a different mechanism that does it more efficiently instead of hacking over an existing field that isn't meant for this. On top of this, there are changes planned for `htlc_maximum_msat`, see https://twitter.com/rusty_twit/status/1532511978006056960 so we won't let node operators freely update this value as it would mess up the results.

**Reply by DerEwige** [reply]
Hmm the twitter only mentions max_htlc_value_in_flight which is the limit for the sum of all HTLC in flight for a given channel. htlc_maximum_msat is the maximum size for one individual HTLC.
Those are different things right? (Granted they may influence each other)

I have actually implemented a dynamic htlc_maximum_msat for my node via plugin and am testing right now. I can write some update once I’ve gathered some data on how it affects my channel, if you are interested.

**Reply by t-bast** [reply]
Woops you're totally right, the planned protocol changes are for `max_htlc_value_in_flight`, not `htlc_maximum_msat`, and they are indeed different things.

The other points still hold though. Also, note that as pm47 mentioned, `htlc_maximum_msat` is static in the current version of the protocol: that means that even if you change the value in a `channel_update`, your node won't actually enforce the new value. But that could still be used as a signal for other nodes, that they can chose to take into account or ignore.

**Reply by DerEwige** [reply]
Agreed, but the current version also mentions:
> SHOULD consider the htlc_maximum_msat when routing.

So even if my node does not enforce it, by anouncing it (not to often) other nodes should consider it when routing and it should lead to better route finding.

I will let it run for a few days then give an update on the result

**Reply by DerEwige** [reply]
@t-bast 
I have been running the dynamic maxHTLC for about a week now.
It works pretty well, but I have one major issue (which is regarding my implementation)

I just could not figure out how to change the state of my local channel (fees) without triggering an announcement to the public.

So right now, I use this to change my fees:
  `  eclair.updateRelayFee(one, new MilliSatoshi(currentBaseFee), currentPercentFee, myTimeout).result(myTimeout.duration(), null);`

This line updates the fees and releases an announcement (with the default maxHTLC)

Then I send another announcement a bit later with my wanted maxHTLCs.
I’m really unhappy with this, because I have to send 2 announcements, when one would be enough and better (because of spam measures)

So if anybody could lend me a hand and show me how to change the fees on my channel without triggering an automated announcement, that would be great.

**Reply by t-bast** [reply]
It doesn't make sense to update fees without telling the network, that's why this triggers the broadcast of a `channel_update`.
You can simply do it in the other order: modify max htlc first, then call `updateRelayFee` which will send the `channel_update` for you.

**Reply by DerEwige** [reply]
That would actually be pretty good.
But right now I do not really change the maxHTLC I just anounce it. (so the actuall enfroces HTLC stayes unchanged and I just annouce a smaller one)
I have the same problem for maxHTLC as for fee. I just could not figure out how to actually change the state in the channel by hand =/

**Reply by t-bast** [reply]
This is protected on purpose, because otherwise there are many ways plugins could completely mess up their channel state and trigger force-closes (or worse, loss of funds). We can't provide unrestricted edit access to the channel state, this is too dangerous.

However we can reasonably provide tools to update specific channel parameters such as max htlc. You can experiment with changing `eclair` to allow this and run your modified version of `eclair` to see if that works for you.

**Reply by DerEwige** [reply]
Ok Thanks,
Will have to push this one back a bit then
